src/server/middleware/transaction/transaction.go

```diff
@@ -0,0 +1,83 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package transaction
+
+import (
+	"context"
+	"errors"
+	"net/http"
+
+	"github.com/goharbor/harbor/src/internal/orm"
+	"github.com/goharbor/harbor/src/server/middleware"
+)
+
+// Config defines the config for Transaction middleware.
+type Config struct {
+	// Acceptor defines a function to commit transaction
+	Acceptor func(statusCode int) bool
+}
+
+var (
+	// DefaultConfig default transaction config
+	DefaultConfig = Config{
+		// acceptor which will commit the transaction when status code between 200 to 308 as success
+		Acceptor: func(statusCode int) bool {
+			// statusCode between 200 to 308 as success
+			return statusCode >= http.StatusOK && statusCode < http.StatusBadRequest
+		},
+	}
+
+	errNonSuccess = errors.New("non success status code")
+)
+
+// Middleware middleware which add transaction for the http request with default config
+func Middleware(skippers ...middleware.Skipper) func(http.Handler) http.Handler {
+	return MiddlewareWithConfig(DefaultConfig, skippers...)
+}
+
+// MiddlewareWithConfig middleware which add transaction for the http request with config
+func MiddlewareWithConfig(config Config, skippers ...middleware.Skipper) func(http.Handler) http.Handler {
+	if config.Acceptor == nil {
+		config.Acceptor = DefaultConfig.Acceptor
+	}
+
+	return middleware.New(func(w http.ResponseWriter, r *http.Request, next http.Handler) {
+		h := func(ctx context.Context) error {
+			rec := &statusRecorder{ResponseWriter: w}
+			next.ServeHTTP(rec, r)
+
+			if !config.Acceptor(rec.statusCode) {
+				return errNonSuccess
+			}
+
+			return nil
+		}
+
+		if err := orm.WithTransaction(h)(r.Context()); err != nil && err != errNonSuccess {
+			// 500 response for db error
+			w.WriteHeader(http.StatusInternalServerError)
+		}
+	}, skippers...)
+}
+
+type statusRecorder struct {
```

Comment:
Use `httptest.ResponseRecorder` instead, there are some constraints for the methods of `http.ResponseWriter` that `statusRecorder` doesn't meet

Reply:
`httptest.ResponseRecorder` use `bytes.Buffer` to store the body content, when the status is success we need to copy the content from `bytes.Buffer` to `http.ResponseWriter`.

This is ok for API's response because its size is not very large, but the response of fetch blob API in v2 may be very very large, the copy may degrade the performance.

Reply:
Understood. The `httptest.ResponseRcorder` isn't the correct solution.
But your implement must meet the constraint of the `http.ResponseWriter` interface:
Should rewrite the `Write` method as it will set the HTTP status code as well?
Make sure the HTTP status  code can only be set once.

How about implement it in the `internal` package and make it as a util so that others can use it if needed?

Reply:
According to the comments of `Write` method in the `http.ResponseWriter`, it writes the data to the connection as part of an HTTP reply. If `WriteHeader` has not yet been called, Write calls WriteHeader(http.StatusOK) before writing the data.

https://github.com/golang/go/blob/master/src/net/http/server.go#L118-L121

It is ok for statusRecorder when we only need to get the status code of the response.

Reply:
Changed to use ResponseBuffer.